src/snowflake/cli/api/project/definition_conversion.py

```diff
@@ -380,33 +450,119 @@ def _convert_templates_in_files(
+        #          }
+        #     }
+        # }
+        # We only convert files on-disk if the "templates" processor is used in the artifacts
+        # and if we're doing a permanent conversion. If we're doing an in-memory conversion,
+        # the CLI global template context is already populated with the v1 definition, so
+        # we don't want to convert the v1 template references in artifact files
+        metrics.set_counter_default(CLICounterField.TEMPLATES_PROCESSOR, 0)
+        artifacts_to_template = [
+            artifact
+            for artifact in pkg_model.artifacts
+            for processor in artifact.processors
+            if processor.name == TEMPLATES_PROCESSOR
+        ]
+        if not in_memory and artifacts_to_template:
+            metrics.set_counter(CLICounterField.TEMPLATES_PROCESSOR, 1)
+
+            # Create a temporary directory to hold the expanded templates,
+            # as if a bundle step had been run but without affecting any
+            # files on disk outside of the artifacts we want to convert
+            with tempfile.TemporaryDirectory() as d:
+                deploy_root = Path(d)
+                bundle_ctx = BundleContext(
+                    package_name=pkg_model.identifier,
+                    artifacts=pkg_model.artifacts,
+                    project_root=project_root,
+                    bundle_root=project_root / pkg_model.bundle_root,
+                    deploy_root=deploy_root,
+                    generated_root=(
+                        project_root / deploy_root / pkg_model.generated_root
+                    ),
+                )
+                template_processor = TemplatesProcessor(bundle_ctx)
+                for artifact in artifacts_to_template:
+                    bundle_map = BundleMap(
+                        project_root=project_root, deploy_root=deploy_root
+                    )
+                    bundle_map.add(artifact)
+
+                    for src, dest in bundle_map.all_mappings(
+                        absolute=True, expand_directories=False
+                    ):
+                        # Copy the mapping from the source root to the deploy root,
+                        # since the processor expects the artifacts to have
+                        # already been bundled (we can't call build_bundle() since it
+                        # checks that the deploy_root is a child of the project_root,
+                        # which isn't the case here)
+                        symlink_or_copy(src, dest, deploy_root=deploy_root)
+
+                    for src, dest in bundle_map.all_mappings(
+                        absolute=True, expand_directories=True
+                    ):
+                        if src.is_dir():
+                            continue
+                        # We call the implementation directly instead of calling process()
+                        # since we need access to the BundleMap to copy files anyways
```

Comment:
it's not really because of the directory, it's because the implementation of `TemplatesProcessor.process()` is just
```python
        get_cli_context().metrics.set_counter(CLICounterField.TEMPLATES_PROCESSOR, 1)

        bundle_map = BundleMap(
            project_root=self._bundle_ctx.project_root,
            deploy_root=self._bundle_ctx.deploy_root,
        )
        bundle_map.add(artifact_to_process)

        for src, dest in bundle_map.all_mappings(
            absolute=True,
            expand_directories=True,
        ):
            self.expand_templates_in_file(src, dest)
```
and since we need access to the `BundleMap` to copy the files back to the source directory, it's simpler to just copy the implementation and add the `copytree` call in the same `for` loop. If we call `.process()`, we'd still have to instantiate the same `BundleMap` and iterate over it again to copy the edited files back